Clang build support

robambalu opened the pull request.

AdamGlustein reviewed (approved):
LGTM

ngoldbaum commented:
We should bring the github actions build changes over from #83, we can also update the build instructions to no longer reference installing g++ or using it on MacOS.

timkpaine commented:
This code does not work on Mac Clang. The first of many, many errors (which do not occur in #83)

```
/Users/timkpaine/Developer/projects/point72/csp/cpp/csp/engine/Struct.h:345:9: error: static assertion failed due to requirement 'std::is_same<std::__bit_const_reference<std::vector<bool, std::allocator<bool>>>, bool>::value || std::is_same<std::vector<std::__bit_const_reference<std::vector<bool, std::allocator<bool>>>, std::allocator<std::__bit_const_reference<std::vector<bool, std::allocator<bool>>>>>, bool>::value'
        static_assert(std::is_same<V,ElemT>::value || std::is_same<std::vector<V>,ElemT>::value );
```

svatasoiu commented on cpp/csp/engine/Struct.h:
```diff
@@ -368,7 +369,7 @@ class ArrayStructField : public NonNativeStructField
     void deepcopyFrom( const Struct * src, Struct * dest ) const override
     {
-        deepcopy( value( src ), value( dest ) );
+        //deepcopy( value( src ), value( dest ) );
```
is this intentional?

robambalu replied:
Accientally commited this change along with the merge since i fixed a conflict here.
This PR is still WIP, not ready for review

ngoldbaum commented:
The last push adds support and tests for mac conda builds, updates the docs, and removes some remanants of using gcc in the mac builds. I don't know if there are additional fixes needed, I'll comment again once I'm done iterating on this.

ngoldbaum commented:
It turns out building using the conda compilers on an intel mac requires setting the min required SDK version to 10.12 (conda-forge defaults to 10.9). I'm not sure how to do that on the github actions runner using a regular `pip install` invocation outside of the `conda build` infrastructure, so I'm punting on testing conda-based builds on intel macs for now. Otherwise this should be good to go as far as the build goes, I think. Maybe we should delete the commented code for supporting cross-compiling on MacOS now arm64 runners are available and it's not needed on github actions?

I think the `ld_classic` thing is a workaround for a MacOS bug that is fixed in the latest MacOS builds and can also be safely deleted.

ngoldbaum commented:
Hmm, not sure what's up with this:

```
=================================== FAILURES ===================================
_________________________ TestEngine.test_threaded_run _________________________

self = <csp.tests.test_engine.TestEngine testMethod=test_threaded_run>

    def test_threaded_run(self):
        # simple test
        runner = csp.run_on_thread(
            csp.count, csp.timer(timedelta(seconds=1)), starttime=datetime(2021, 4, 23), endtime=timedelta(seconds=60)
        )
        res = runner.join()[0]
        self.assertEqual(len(res), 60)
        # ensure stopping doesnt try to access dead push input adapter
        runner.stop_engine()
    
        # realtime
        @csp.graph
        def g(count: int) -> csp.ts[int]:
            x = csp.count(csp.timer(timedelta(seconds=0.1)))
            stop = x == count
            stop = csp.filter(stop, stop)
    
            csp.stop_engine(stop)
            return x
    
        runner = csp.run_on_thread(g, 5, starttime=datetime.utcnow(), endtime=timedelta(seconds=60), realtime=True)
        res = runner.join()[0]
        self.assertEqual(len(res), 5)
    
        # midway stop
        runner = csp.run_on_thread(g, 50000, starttime=datetime.utcnow(), endtime=timedelta(minutes=30), realtime=True)
        import time
    
        time.sleep(1)
        runner.stop_engine()
        res = runner.join()[0]
>       self.assertLess(len(res), 1000)
E       AssertionError: 17999 not less than 1000

csp/tests/test_engine.py:1406: AssertionError
```

Only happens using the conda build environment on an ARM mac. I don't reproduce this failure on my M3 macbook pro using a conda-based build, so not sure why this is a problem on github actions.

ngoldbaum commented:
@isuruf pointed out the remnants of the cross-compilation support was breaking the build for intel macs, so I've added back the intel mac tests.

Still not sure what's up with the test failure on ARM macs. I reduced the timeout so at least we don't have to wait 30 minutes for the test to fail...

robambalu commented:
> Hmm, not sure what's up with this:
> 
> ```
> =================================== FAILURES ===================================
> _________________________ TestEngine.test_threaded_run _________________________
> 
> self = <csp.tests.test_engine.TestEngine testMethod=test_threaded_run>
> 
>     def test_threaded_run(self):
>         # simple test
>         runner = csp.run_on_thread(
>             csp.count, csp.timer(timedelta(seconds=1)), starttime=datetime(2021, 4, 23), endtime=timedelta(seconds=60)
>         )
>         res = runner.join()[0]
>         self.assertEqual(len(res), 60)
>         # ensure stopping doesnt try to access dead push input adapter
>         runner.stop_engine()
>     
>         # realtime
>         @csp.graph
>         def g(count: int) -> csp.ts[int]:
>             x = csp.count(csp.timer(timedelta(seconds=0.1)))
>             stop = x == count
>             stop = csp.filter(stop, stop)
>     
>             csp.stop_engine(stop)
>             return x
>     
>         runner = csp.run_on_thread(g, 5, starttime=datetime.utcnow(), endtime=timedelta(seconds=60), realtime=True)
>         res = runner.join()[0]
>         self.assertEqual(len(res), 5)
>     
>         # midway stop
>         runner = csp.run_on_thread(g, 50000, starttime=datetime.utcnow(), endtime=timedelta(minutes=30), realtime=True)
>         import time
>     
>         time.sleep(1)
>         runner.stop_engine()
>         res = runner.join()[0]
> >       self.assertLess(len(res), 1000)
> E       AssertionError: 17999 not less than 1000
> 
> csp/tests/test_engine.py:1406: AssertionError
> ```
> 
> Only happens using the conda build environment on an ARM mac. I don't reproduce this failure on my M3 macbook pro using a conda-based build, so not sure why this is a problem on github actions.

It does look like this is pointing to some strange issue.
This realtime run should be ticking once every 100 ms.  We launch it on a separate thread, sleep for 1 second ( should be around 10 ticks ) then assert the count is less than *1000* for extreme leeway.
In this test it somehow got 18000 ticks.
So that means it ran for the fulll 30 minutes, but the whole point of this test is to ensure that runner.stop_engine() Call stops it mid run after 1 second.  So something isnt actually working with stop_engine here

ngoldbaum commented:
I added a step to the end of the action [to trigger a remote debugging session](https://github.com/marketplace/actions/debugging-with-tmate) and managed to reproduce this issue. It is a deadlock, here's the traceback from both threads:

```
Thread 0x000000016f90b000 (most recent call first):
  File "/Users/runner/work/csp/csp/csp/impl/wiring/runtime.py", line 125 in _run_engine
  File "/Users/runner/work/csp/csp/csp/impl/wiring/runtime.py", line 227 in run
  File "/Users/runner/work/csp/csp/csp/impl/wiring/runtime.py", line 241 in run
  File "/Users/runner/work/csp/csp/csp/impl/wiring/threaded_runtime.py", line 81 in _run
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 982 in run
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 1045 in _bootstrap_inner
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 1002 in _bootstrap

Thread 0x00000001e0325000 (most recent call first):
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 1139 in _wait_for_tstate_
lock
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 1119 in join
  File "/Users/runner/work/csp/csp/csp/impl/wiring/threaded_runtime.py", line 92 in join
  File "/Users/runner/work/csp/csp/csp/tests/test_engine.py", line 1405 in test_threaded_run
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/unittest/case.py", line 579 in _callTestMethod
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/unittest/case.py", line 623 in run
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/unittest/case.py", line 678 in __call__
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/_pytest/unittest.py", line 321 in runtest
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/_pytest/runner.py", line 172 in
 pytest_runtest_call
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/pluggy/_callers.py", line 102 i
n _multicall
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/pluggy/_manager.py", line 119 i
n _hookexec
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/pluggy/_hooks.py", line 501 in
__call__
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/_pytest/runner.py", line 240 in
 <lambda>
  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/_pytest/runner.py", line 340
<further frames elided>
```


So it looks like for whatever reason `engine.run()` on the work thread never returns.

Happy to do further debugging here or walk someone else through how to set this up.

robambalu commented:
Doesn’t look like its deadlocked here per se, just that the engine never got notified to exit
If you change the endtime of this specific run to be say 1 minute instead of 30, I expect it will exit after 1 minute ( not stay deadlocked )
The question is more, why doesn’t the engine get notified of this.
Have you figured out any way to reproduce this locally?  If I run on my macbook it seems to pass fine.
If its only remote and you can get some log output, adding some prints may help diagnose this too. I would put a print at:
the start of the g() graph
before and after the sleep call
in _EngineStopper __init__ and register_impl
and in _EngineStopper stop_engine, also printing self._pushimpl

From: Nathan Goldbaum ***@***.***>
Sent: Wednesday, March 20, 2024 11:08 AM
To: Point72/csp ***@***.***>
Cc: Ambalu, Robert ***@***.***>; Author ***@***.***>
Subject: Re: [Point72/csp] Clang build support (PR #132)

I added a step to the end of the action to trigger a remote debugging session and managed to reproduce this issue. It is a deadlock, here's the traceback from both threads: Thread 0x000000016f90b000 (most recent call first): File "/Users/runner/work/csp/csp/csp/impl/wiring/runtime. py",


I added a step to the end of the action to trigger a remote debugging session<https://urldefense.com/v3/__https:/github.com/marketplace/actions/debugging-with-tmate__;!!K_TC0FI_KA!r-SgRQWoNBdYNo6QZSvm98aynG7bDWgTey5wTK66-qlz-99xmZPOVjIE9uVXEwR6Qdo5trx8emxHmtbHs5iZDC7VNJweiA$> and managed to reproduce this issue. It is a deadlock, here's the traceback from both threads:

Thread 0x000000016f90b000 (most recent call first):

  File "/Users/runner/work/csp/csp/csp/impl/wiring/runtime.py", line 125 in _run_engine

  File "/Users/runner/work/csp/csp/csp/impl/wiring/runtime.py", line 227 in run

  File "/Users/runner/work/csp/csp/csp/impl/wiring/runtime.py", line 241 in run

  File "/Users/runner/work/csp/csp/csp/impl/wiring/threaded_runtime.py", line 81 in _run

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 982 in run

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 1045 in _bootstrap_inner

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 1002 in _bootstrap



Thread 0x00000001e0325000 (most recent call first):

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 1139 in _wait_for_tstate_

lock

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 1119 in join

  File "/Users/runner/work/csp/csp/csp/impl/wiring/threaded_runtime.py", line 92 in join

  File "/Users/runner/work/csp/csp/csp/tests/test_engine.py", line 1405 in test_threaded_run

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/unittest/case.py", line 579 in _callTestMethod

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/unittest/case.py", line 623 in run

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/unittest/case.py", line 678 in __call__

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/_pytest/unittest.py", line 321 in runtest

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/_pytest/runner.py", line 172 in

 pytest_runtest_call

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/pluggy/_callers.py", line 102 i

n _multicall

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/pluggy/_manager.py", line 119 i

n _hookexec

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/pluggy/_hooks.py", line 501 in

__call__

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/_pytest/runner.py", line 240 in

 <lambda>

  File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/_pytest/runner.py", line 340

<further frames elided>

So it looks like for whatever reason engine.run() on the work thread never returns.

Happy to do further debugging here or walk someone else through how to set this up.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/Point72/csp/pull/132*issuecomment-2009799247__;Iw!!K_TC0FI_KA!r-SgRQWoNBdYNo6QZSvm98aynG7bDWgTey5wTK66-qlz-99xmZPOVjIE9uVXEwR6Qdo5trx8emxHmtbHs5iZDC6rplqoAg$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AG22D7DYTPJMNBEX2YWAC33YZGQ4VAVCNFSM6AAAAABEF7VO46VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMBZG44TSMRUG4__;!!K_TC0FI_KA!r-SgRQWoNBdYNo6QZSvm98aynG7bDWgTey5wTK66-qlz-99xmZPOVjIE9uVXEwR6Qdo5trx8emxHmtbHs5iZDC5EbSzr0w$>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>





DISCLAIMER: This e-mail message and any attachments are intended solely for the use of the individual or entity to which it is addressed and may contain information that is sensitive, confidential, legally privileged or otherwise protected under applicable law. No confidentiality or privilege is waived or lost by transmission error. If you are not the intended recipient, you are hereby notified that any disclosure, dissemination, distribution, copying or other use of this message or its attachments is strictly prohibited. If you have received this message in error, please notify the sender immediately and permanently delete this message and any attachments.

Point72 reserves the right, to the extent required and/or permitted under applicable law, to monitor electronic communications, including telephone calls with Point72 personnel. This message is subject to the Point72 Terms of Use (https://point72.com/terms-of-use/) and our Global Privacy Policy (https://point72.com/global-privacy-policy), which describe your rights and how we handle personal and other data. If you cannot access these links, please notify us by replying to this message and we will send the contents to you.

Point72 does not warrant this e-mail message or any attachments is virus-free and is not liable for any losses or responsibilities arising from any virus being transmitted by this communication. E-mail communication is not a secure method of communication, can be intercepted and cannot be guaranteed to be error free.

By communicating with Point72 you acknowledge that you have read, understood and consented, where legally required and permitted, to the Disclaimer above, including the Point72 Terms of Use and the Global Privacy Policy referenced therein.

ngoldbaum commented:
> Have you figured out any way to reproduce this locally?  If I run on my macbook it seems to pass fine.

Nope, I've only seen this behavior on the github actions runner 😢 

> adding some prints may help diagnose this too

I added some prints in [this commit](https://github.com/ngoldbaum/csp/commit/e3be31f113761fc0600c9ec8e7ff84ed39e335a1). You can see a failing run with those prints [here](https://github.com/ngoldbaum/csp/actions/runs/8362385698/job/22892840442#step:7:904) - deeplinking into the log where the prints start.

So it looks like `runner.stop_engine()` does get called, and then a hang happens in the C++ code somewhere.

I added an `fprintf` call inside `PyPushInputAdapter.pushPyTick`:

```diff
diff --git a/cpp/csp/python/PyPushInputAdapter.cpp b/cpp/csp/python/PyPushInputAdapter.cpp
index 3b5f91b..26a9892 100644
--- a/cpp/csp/python/PyPushInputAdapter.cpp
+++ b/cpp/csp/python/PyPushInputAdapter.cpp
@@ -156,6 +156,7 @@ public:
         {
             if( !validatePyType( this -> dataType(), m_pyType.ptr(), value ) )
                 CSP_THROW( TypeError, "" );
+            fprintf(stderr, "PyPushInputAdapter.pushPyTick called\n");
             pushTick<T>( std::move(fromPython<T>( value, *this -> dataType() )), batch );
         }
         catch( const TypeError & )
```

And that seems to make the failures happen less often, but they still happen, and that `fprintf` does get called before the hang. I'm having a little trouble navigating the C++ internals - any other likely spots I can put `fprintf` calls to narrow down where the hang is happening?

timkpaine reviewed (commented):
we may see better behavior after recent merges, I too noticed these random hangs but they seem to have cleared up after applying some of the changes that were just merged.

robambalu commented on CMakeLists.txt:
```diff
@@ -170,7 +160,7 @@ endif()
-    set(CMAKE_INSTALL_RPATH "@loader_path/:@loader_path/../lib")
+    set(CMAKE_INSTALL_RPATH "@loader_path/")
 elseif(LINUX)
     set(CMAKE_INSTALL_RPATH "\$ORIGIN:\$ORIGIN/../lib")
```
trying

robambalu commented:
> > Have you figured out any way to reproduce this locally?  If I run on my macbook it seems to pass fine.
> 
> Nope, I've only seen this behavior on the github actions runner 😢
> 
> > adding some prints may help diagnose this too
> 
> I added some prints in [this commit](https://github.com/ngoldbaum/csp/commit/e3be31f113761fc0600c9ec8e7ff84ed39e335a1). You can see a failing run with those prints [here](https://github.com/ngoldbaum/csp/actions/runs/8362385698/job/22892840442#step:7:904) - deeplinking into the log where the prints start.
> 
> So it looks like `runner.stop_engine()` does get called, and then a hang happens in the C++ code somewhere.
> 
> I added an `fprintf` call inside `PyPushInputAdapter.pushPyTick`:
> 
> ```diff
> diff --git a/cpp/csp/python/PyPushInputAdapter.cpp b/cpp/csp/python/PyPushInputAdapter.cpp
> index 3b5f91b..26a9892 100644
> --- a/cpp/csp/python/PyPushInputAdapter.cpp
> +++ b/cpp/csp/python/PyPushInputAdapter.cpp
> @@ -156,6 +156,7 @@ public:
>          {
>              if( !validatePyType( this -> dataType(), m_pyType.ptr(), value ) )
>                  CSP_THROW( TypeError, "" );
> +            fprintf(stderr, "PyPushInputAdapter.pushPyTick called\n");
>              pushTick<T>( std::move(fromPython<T>( value, *this -> dataType() )), batch );
>          }
>          catch( const TypeError & )
> ```
> 
> And that seems to make the failures happen less often, but they still happen, and that `fprintf` does get called before the hang. I'm having a little trouble navigating the C++ internals - any other likely spots I can put `fprintf` calls to narrow down where the hang is happening?

@ngoldbaum that pushTick method should never block, so I assume its getting past that.  It somewhat feels like the python gil may not be getting released at all or enough.  Its really hard to debug without being able to reproduce outside of the github workflows.  I'm going to try adding logs now as well.

robambalu commented:
> I added a step to the end of the action [to trigger a remote debugging session](https://github.com/marketplace/actions/debugging-with-tmate) and managed to reproduce this issue. It is a deadlock, here's the traceback from both threads:
> 
> ```
> Thread 0x000000016f90b000 (most recent call first):
>   File "/Users/runner/work/csp/csp/csp/impl/wiring/runtime.py", line 125 in _run_engine
>   File "/Users/runner/work/csp/csp/csp/impl/wiring/runtime.py", line 227 in run
>   File "/Users/runner/work/csp/csp/csp/impl/wiring/runtime.py", line 241 in run
>   File "/Users/runner/work/csp/csp/csp/impl/wiring/threaded_runtime.py", line 81 in _run
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 982 in run
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 1045 in _bootstrap_inner
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 1002 in _bootstrap
> 
> Thread 0x00000001e0325000 (most recent call first):
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 1139 in _wait_for_tstate_
> lock
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/threading.py", line 1119 in join
>   File "/Users/runner/work/csp/csp/csp/impl/wiring/threaded_runtime.py", line 92 in join
>   File "/Users/runner/work/csp/csp/csp/tests/test_engine.py", line 1405 in test_threaded_run
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/unittest/case.py", line 579 in _callTestMethod
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/unittest/case.py", line 623 in run
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/unittest/case.py", line 678 in __call__
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/_pytest/unittest.py", line 321 in runtest
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/_pytest/runner.py", line 172 in
>  pytest_runtest_call
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/pluggy/_callers.py", line 102 i
> n _multicall
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/pluggy/_manager.py", line 119 i
> n _hookexec
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/pluggy/_hooks.py", line 501 in
> __call__
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/_pytest/runner.py", line 240 in
>  <lambda>
>   File "/Users/runner/micromamba/envs/csp-dev/lib/python3.11/site-packages/_pytest/runner.py", line 340
> <further frames elided>
> ```
> 
> So it looks like for whatever reason `engine.run()` on the work thread never returns.
> 
> Happy to do further debugging here or walk someone else through how to set this up.

If you can show me how to set this up that would be great.  Do you get a remote gdb session or pdb?  I would likely need gdb for this, or the macos equivalent

timkpaine commented on cpp/tests/core/test_time.cpp:
```diff
@@ -159,6 +159,7 @@ TEST( DateTimeEx, test_basic_functionality )
     ASSERT_EQ( ex2.microseconds(), 123456 );
     ASSERT_EQ( ex2.nanoseconds(),  123456789 );
 
+#ifdef __linux__
```
i have code for this somewhere, let me find it

robambalu replied:
its in the windows build branch, this should cover both

robambalu replied:
im ok skipping this test if !linux

timkpaine commented:
full build: https://github.com/Point72/csp/actions/runs/8452535222

timkpaine commented:
full build isnt running mac tests, testing in #168

timkpaine commented:
actual full build this time (+ mac tests!) https://github.com/Point72/csp/actions/runs/8455095906

robambalu commented:
curious why are we still using vcpkg for the full build if the conda builds are working ( and are significantly faster )

ngoldbaum commented:
The wheels need to be built using the toolchain on the manylinux image, so something like vcpkg is necessary. You can’t use the toolchain conda-forge is built with.

timkpaine commented:
gtg 
![image](https://github.com/Point72/csp/assets/3105306/96bb01bf-b938-483c-8264-bdf9d1152926)

timkpaine reviewed (approved):
lgtm
